### ISSUE TITLE
Bookmark activities feature

### ISSUE THREAD

**Issue by bytech-ca** [eclectic-bytech/webcourses-ngx-v4#275]
- [x] wngx_bookmarks table with the following fields: id(auto_increment), pid(fk on user progress table), aid(fk on course_syllabus), label(varchar 256 length)
- [x] font awesome bookmark icon at top of page, between sidebar and 'make this activity part of the demo'
- [x] bookmark icon and label in sidebar: Progress, Index, Publisher, Help, Bookmarks (partial code is already included in the template, but commented out)
- [x] laravel end point `/v4/webcourse/activities/bookmark/{aid}`

Please add brief descriptions of each commit - it'll help me review the code more efficiently. I have no idea what some bits of the code are doing, and it will be helpful to be able to check the commit comment and get extra insight on what a commits purpose is.

Some reviews are also getting held up because unrelated changes "that just made sense to include" need updates, and there is no intuitive way for me to comment on these without a proper issue... things get bogged down.

**Comment by bytech-ca** [comment]
For the Laravel routes, and having just one, I wonder if something like this will work:

```
            Route::get('/bookmark/{id}', [BookmarkController::class, 'add_bookmark']);
            Route::delete('/bookmark/{id}', [BookmarkController::class, 'delete_bookmark']);
```

**Comment by bytech-ca** [comment]
Also found this on ChatGPT

In Laravel, you can run different methods in a controller depending on the HTTP request type (e.g., GET, POST, PUT, DELETE) using conditional statements. Laravel provides convenient methods for handling different HTTP verbs in controllers. Here's how you can do it:

1. Define the Routes:
First, you need to define the routes in your `routes/web.php` or `routes/api.php` file. For example:

```php
Route::get('/example', 'ExampleController@index'); // Handles GET request
Route::post('/example', 'ExampleController@store'); // Handles POST request
Route::put('/example/{id}', 'ExampleController@update'); // Handles PUT request
Route::delete('/example/{id}', 'ExampleController@destroy'); // Handles DELETE request
```

2. Create the Controller:
Create a new controller or use an existing one using the `make:controller` Artisan command:

```bash
php artisan make:controller ExampleController
```

3. Implement the Controller:
Inside your `ExampleController.php`, you can define different methods to handle each HTTP request type:

```php
namespace App\Http\Controllers;

use Illuminate\Http\Request;

class ExampleController extends Controller
{
    public function index()
    {
        // Logic for handling GET request
        return response()->json(['message' => 'This is a GET request']);
    }

    public function store(Request $request)
    {
        // Logic for handling POST request
        return response()->json(['message' => 'This is a POST request']);
    }

    public function update(Request $request, $id)
    {
        // Logic for handling PUT request
        return response()->json(['message' => "This is a PUT request for item $id"]);
    }

    public function destroy($id)
    {
        // Logic for handling DELETE request
        return response()->json(['message' => "This is a DELETE request for item $id"]);
    }
}
```

In this example, the `index()` method is executed when a GET request is made to `/example`, the `store()` method is executed for a POST request to the same URL, the `update()` method is executed for a PUT request with a dynamic `$id` parameter, and the `destroy()` method is executed for a DELETE request with a dynamic `$id` parameter.

Now, when you make a specific HTTP request to the corresponding route, Laravel will automatically call the appropriate method in the controller based on the HTTP verb used.

**Comment by bytech-ca** [comment]
And another ChatGPT solution that might work, but looks a bit less clean...

In Laravel, you can define a route that handles both GET and POST requests using the `match` method or the `any` method. These methods allow you to specify multiple HTTP verbs for a single route. Here's how you can set up a route that handles both GET and POST requests using the same route:

Using `match` method:

```php
use Illuminate\Support\Facades\Route;

Route::match(['get', 'post'], '/your-route', function () {
    // Your logic here for handling both GET and POST requests
    // You can use the request method to differentiate between the two if needed
    if (request()->isMethod('post')) {
        // Handle POST request
    } else {
        // Handle GET request
    }
});
```

Using `any` method:

```php
use Illuminate\Support\Facades\Route;

Route::any('/your-route', function () {
    // Your logic here for handling both GET and POST requests
    // You can use the request method to differentiate between the two if needed
    if (request()->isMethod('post')) {
        // Handle POST request
    } else {
        // Handle GET request
    }
});
```

Both approaches will allow the same route to handle both GET and POST requests. In the route closure or controller method, you can check the HTTP method (`GET` or `POST`) using the `request()->isMethod()` method and then perform the specific actions accordingly.

**Comment by SirL3X0N** [comment]
> For the Laravel routes, and having just one, I wonder if something like this will work:
> 
> ```
>             Route::get('/bookmark/{id}', [BookmarkController::class, 'add_bookmark']);
>             Route::delete('/bookmark/{id}', [BookmarkController::class, 'delete_bookmark']);
> ```


I thought you wanted the links merged into one but this would've been so easy

**Comment by bytech-ca** [comment]
A typo in the video: `[New Link]` should be `[New Line]`

https://github.com/eclectic-bytech/webcourses-ngx-v4/assets/33671154/f3a9b01e-02cc-428b-a95a-7af36d3f2f6a

**Comment by bytech-ca** [comment]
Found why it's not working for me... you need to download and import the "real sql data" from CL repo: bookmarking only works on course ID 1.

I made some fixes and uploaded the code. Haven't done any testing after spotting the error... please

- [ ] test with real data 
- [ ] correct grammar
- [ ] add padding / styling consistent with other sidebar tabs

**Comment by bytech-ca** [comment]
Updated and merged into master 26afd3460eb6865f70befb6266feeee6e680790f
